### PR TITLE
Update BLST And Herumi

### PR DESCRIPTION
**Pull request by nisdas**
**What type of PR is this?**

Bug Fix

**What does this PR do? Why is it needed?**

- [x] In #7537, our bls go wrapper was updated but neither of the source libraries were updated which caused a mismatch 
when compiling from source. This PR fixes that by updating all the relevant dependencies and cleaning up mcl builds.
- [x] Update Blst and Herumi deps
- [x] Clean Up BLS methods to allow for sharing common libraries.

**Which issues(s) does this PR fix?**

N.A

**Other notes for review**

### CODE REVIEW

**Review by shayzluf — COMMENTED** [review]
need to update the comment as we don't use the random any more

// VerifyMultipleSignatures verifies a non-singular set of signatures and its respective pubkeys and messages.
// This method provides a safe way to verify multiple signatures at once. We pick a number randomly from 1 to max
// uint64 and then multiply the signature by it. We continue doing this for all signatures and its respective pubkeys.
// S* = S_1 * r_1 + S_2 * r_2 + ... + S_n * r_n
// P'_{i,j} = P_{i,j} * r_i
// e(S*, G) = \prod_{i=1}^n \prod_{j=1}^{m_i} e(P'_{i,j}, M_{i,j})
// Using this we can verify multiple signatures safely.

File: shared/bls/common/constants.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+// SecretKeyIsZero checks the validity of a secret key.
+func SecretKeyIsZero(key []byte) bool {
+	return bytes.Equal(key, ZeroSecretKey[:])
+}
+
+// PublicKeyIsInfinite checks the validity of a public key.
+func PublicKeyIsInfinite(key []byte) bool {
+	return bytes.Equal(key, InfinitePublicKey[:])
+}
```
Comment (prestonvanloon) [inline shared/bls/common/constants.go:22]: Is there a faster way to do this based on each library? I think marshalling to bytes is pretty slow, but it looks like you can complete this check in herumi like `pubKey.(*PublicKey).p.IsZero()`. Is that the same?

If so, you can accept some pubkey interface and call IsInfinite where the underlying library has support to detect this without marshalling to bytes. 

Same feedback for each of these three methods.

File: shared/bls/spectest/aggregate_verify_test.go [hunk 2]
```diff
@@ -36,13 +35,18 @@ func testAggregateVerifyYaml(t *testing.T) {
...
-			pubkeys := make([]iface.PublicKey, 0, len(test.Input.Pubkeys))
+			pubkeys := make([]common.PublicKey, 0, len(test.Input.Pubkeys))
 			msgs := make([][32]byte, 0, len(test.Input.Messages))
 			for _, pubKey := range test.Input.Pubkeys {
 				pkBytes, err := hex.DecodeString(pubKey[2:])
 				require.NoError(t, err)
 				pk, err := bls.PublicKeyFromBytes(pkBytes)
-				require.NoError(t, err)
+				if err != nil {
+					if test.Output == false && err == common.ErrInfinitePubKey {
```
Comment (prestonvanloon) [inline shared/bls/spectest/aggregate_verify_test.go:45]:
```suggestion
					if test.Output == false && errors.Is(err, common.ErrInfinitePubKey) {
```
I think errors.Is is the preferred way to check errors now.

File: shared/bls/spectest/fast_aggregate_verify_test.go [hunk 3]
```diff
@@ -37,16 +37,25 @@ func testFastAggregateVerifyYaml(t *testing.T) {
...
 
-			pubkeys := make([]iface.PublicKey, len(test.Input.Pubkeys))
+			pubkeys := make([]common.PublicKey, len(test.Input.Pubkeys))
 			for j, raw := range test.Input.Pubkeys {
 				pkBytes, err := hex.DecodeString(raw[2:])
 				require.NoError(t, err)
 				pk, err := bls.PublicKeyFromBytes(pkBytes)
-				require.NoError(t, err)
+				if err != nil {
+					if test.Output == false && err == common.ErrInfinitePubKey {
```
Comment (prestonvanloon) [inline shared/bls/spectest/fast_aggregate_verify_test.go:46]:
```suggestion
					if test.Output == false && errors.Is(err, common.ErrInfinitePubKey) {
```

File: shared/bls/spectest/fast_aggregate_verify_test.go [hunk 4]
```diff
@@ -37,16 +37,25 @@ func testFastAggregateVerifyYaml(t *testing.T) {
...
+				}
 				pubkeys[j] = pk
 			}
 
-			msgBytes, err := hex.DecodeString(test.Input.Message[2:])
+			msg := test.Input.Message
+			if msg == "" {
+				msg = test.Input.Messages
+			}
+			msgBytes, err := hex.DecodeString(msg[2:])
```
Comment (prestonvanloon) [inline shared/bls/spectest/fast_aggregate_verify_test.go:58]: Please add a comment or TODO why we have to do this. It should be fixed soon with https://github.com/ethereum/eth2.0-spec-tests/issues/22

File: shared/bls/spectest/sign_test.go [hunk 5]
```diff
@@ -40,8 +40,12 @@ func testSignMessageYaml(t *testing.T) {
 			pkBytes, err := hex.DecodeString(test.Input.Privkey[2:])
 			require.NoError(t, err)
 			sk, err := bls.SecretKeyFromBytes(pkBytes)
-			require.NoError(t, err)
-
+			if err != nil {
+				if test.Output == "" && err == common.ErrZeroKey {
```
Comment (prestonvanloon) [inline shared/bls/spectest/sign_test.go:44]:
```suggestion
				if test.Output == "" && errors.Is(err, common.ErrZeroKey) {
```

File: shared/bls/spectest/verify_test.go [hunk 6]
```diff
@@ -38,8 +39,12 @@ func testVerifyMessageYaml(t *testing.T) {
 			pkBytes, err := hex.DecodeString(test.Input.Pubkey[2:])
 			require.NoError(t, err)
 			pk, err := bls.PublicKeyFromBytes(pkBytes)
-			require.NoError(t, err)
-
+			if err != nil {
+				if test.Output == false && err == common.ErrInfinitePubKey {
```
Comment (prestonvanloon) [inline shared/bls/spectest/verify_test.go:43]:
```suggestion
				if test.Output == false && errors.Is(err, common.ErrInfinitePubKey) {
```

File: third_party/herumi/mcl.BUILD [hunk 7]
```diff
@@ -66,6 +65,8 @@ cc_library(
         "include/mcl/lagrange.hpp",
         "include/mcl/bn.hpp",
         "include/mcl/operator.hpp",
+        "include/mcl/window_method.hpp",
+
```
Comment (prestonvanloon) [inline third_party/herumi/mcl.BUILD:69]:
```suggestion
```

File: shared/bls/spectest/fast_aggregate_verify_test.yaml.go [hunk 8]
```diff
@@ -7,6 +7,7 @@ type FastAggregateVerifyTest struct {
 	Input struct {
 		Pubkeys   []string `json:"pubkeys"`
 		Message   string   `json:"message"`
+		Messages  string   `json:"messages"`
```
Comment (prestonvanloon) [inline shared/bls/spectest/fast_aggregate_verify_test.yaml.go:10]: Same comment here about a TODO or comment. This should be removed after https://github.com/ethereum/eth2.0-spec-tests/issues/22 is resolved.

File: shared/bls/common/error.go [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+import "errors"
+
+// ErrZeroKey describes an error due to a zero secret key.
+var ErrZeroKey = errors.New("received secret key is zero")
+
+// ErrInfinitePubKey describes an error due to an infinite public key.
+var ErrInfinitePubKey = errors.New("received an infinite public key")
+
+// ErrInfiniteSignature describes an error due to an infinite signature.
+var ErrInfiniteSignature = errors.New("received an infinite signature")
```
[the remark this answers is not in the document]
Comment (nisdas) [reply shared/bls/common/error.go:12]: We can't make it as part of `shared/bls` as that would make it a circular dependency. If the errors are in `shared/bls` , and imported into either blst/herumi, the compiler will complain as we import blst/herumi into `bls/shared`. Although you do make a good point on the fact that external consumers might need this. What if I make it a separate package form `iface` ?

File: shared/bls/common/constants.go [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+
+// SignatureIsInfinite checks the validity of a signature.
+func SignatureIsInfinite(key []byte) bool {
+	return bytes.Equal(key, InfiniteSignature[:])
+}
```
[the remark this answers is not in the document]
Comment (nisdas) [reply shared/bls/common/constants.go:27]: I think maybe I can simply add in an additional method to public keys and secret keys and do away with these checks.

File: shared/bls/herumi/public_key.go [hunk 11]
```diff
@@ -3,10 +3,12 @@ package herumi
 import (
 	"fmt"
 
+	"github.com/prysmaticlabs/prysm/shared/bls/iface"
+
```
Comment (nisdas) [inline shared/bls/herumi/public_key.go:7]:
```suggestion

```

File: shared/bls/blst/signature_test.go [hunk 12]
```diff
@@ -9,12 +9,14 @@ import (
 	"testing"
 
 	"github.com/prysmaticlabs/prysm/shared/bls/iface"
+
```
Comment (nisdas) [inline shared/bls/blst/signature_test.go:12]:
```suggestion

```

File: shared/bls/blst/signature.go [hunk 13]
```diff
@@ -6,8 +6,9 @@ package blst
 import (
 	"fmt"
 
-	"github.com/pkg/errors"
 	"github.com/prysmaticlabs/prysm/shared/bls/iface"
+
```
Comment (nisdas) [inline shared/bls/blst/signature.go:10]:
```suggestion

```

**Review by prestonvanloon — APPROVED** [review]

**Review by terencechain — APPROVED** [review]

**Pull request merged**